proposals/CloudIdentifier.md

```diff
@@ -0,0 +1,197 @@
+2. VDI web app sends the file’s cloud identifier to VDI server
+3. VDI server requests and receives the file’s content from CSP server using the cloud identifier
+4. VDI server sends any changes made to the file back to the CSP server
+5. CSP client synchronizes updated file to local device (might be metadata only)
+
+In this scenario only transfers (3) and (4) would actually transfer the full file, all the other transfers only move the cloud identifier. These actual file transfers also don’t move across the device’s network connection, but rather use the VDI server’s network connection to the CSP server, which should likely also have higher bandwidth.
+
+This reduces network traffic and delays (especially with large files) and prevents version drift (file being modified somewhere else while a local client device is uploading).
+
+#### De-duplication for online document editors
+
+Web-based document editors can already open local files using existing file system APIs. In order to offer full functionality, including cross device support and some advanced editing/sharing features, the documents need to be uploaded to a cloud storage (Google Drive or Microsoft OneDrive or similar) before they can be edited. This leads to duplicate files for documents that were already stored in these locations.
+
+With the proposed changes, the web application could check whether the given file handle is already synced by a cloud storage and generate identifiers for these without a duplicate upload.
+
+#### Drag & Drop into Mail
+
+Sharing files via mail used to be done using simple file attachments. Especially for large files, it is preferable to instead upload them to a CSP and then share an access link to that file to the recipient. Google Mail’s web application, for example, already does this by prompting the user to upload a file to Google Drive when attaching a large file.
+
+With the proposed changes, the web application could already detect that the file is synced by cloud storage and generate a share link for that file’s cloud identifier without requiring the user to manually go through these steps or prompt the user to upload the file to cloud storage again.
+
+### Non-Goals
+
+This proposal does not plan to provide a standardized way for CSP’s web APIs to 
+* give access permissions to these files/directories.
+* interact (fetch/modify/etc) with these files/directories.
+* provide additional meta-data (e.g. sync status)
+
+Instead, it should only generate an opaque identifier for a file/directory so that web apps can then interact with these individual CSPs’ web APIs on the same file/directory. Obtaining the required access permissions on that file is delegated to the web app.
+Web apps might only support one specific CSP. When supporting multiple CSPs, each CSP will require their own implementation, although semantically they would likely perform similar steps.
```

Comment:
By this I assume you mean that a web app might only know how to talk to, say, Google Drive, but not OneDrive?

The number of CSPs that a web app knows how to talk with seems like an implementation detail of that app, and not all that relevant from the perspective of the `getCloudIdentifiers()` web API? The web API only cares what CSPs from the local machine have registered themselves with the user agent - not which of of those CSPs the calling web app knows how to talk with

Do you think it could be worth making more explicit that this proposal is really just to allow the user agent to act as a bridge between the local CSP client and the CSP server, and that it _does not_ handle the actual reading/writing of files with the CSP? (as summarized nicely in this comment https://github.com/WICG/file-system-access/issues/358#issuecomment-1522280549)

Reply:
rewritten this part a bit to make it more clear. It's a bit weird calling `getCloudIdentifiers()` and the CSPs' APIs both "web API"s. One is a web API in the browser, while the other one is an API towards the CSP back-end. Any suggestions for better naming here?